src/lib/hosts-file.ts

```diff
@@ -0,0 +1,175 @@
+import fs from 'fs';
+import { platform, tmpdir } from 'os';
+import path from 'path';
+import { promisify } from 'util';
+import sudo from 'sudo-prompt';
+
+const readFile = promisify( fs.readFile );
+const writeFile = promisify( fs.writeFile );
+const sudoExec = promisify( sudo.exec );
+
+// Host file paths for different operating systems
+const HOST_FILES: Record< string, string > = {
+	win32: 'C:\\Windows\\System32\\drivers\\etc\\hosts',
+	darwin: '/etc/hosts',
+	linux: '/etc/hosts',
+};
+
+// Get the appropriate hosts file path for the current platform
+const getHostsFilePath = (): string => {
+	const currentPlatform = platform();
+	const hostsPath = HOST_FILES[ currentPlatform ];
+
+	if ( ! hostsPath ) {
+		throw new Error( `Unsupported platform: ${ currentPlatform }` );
+	}
+
+	return hostsPath;
+};
+
+/**
+ * Reads the hosts file content
+ */
+export const readHostsFile = async (): Promise< string > => {
+	try {
+		const hostsPath = getHostsFilePath();
+		const content = await readFile( hostsPath, 'utf8' );
+		return content;
+	} catch ( error ) {
+		console.error( 'Error reading hosts file:', error );
+		throw error;
+	}
+};
+
+/**
+ * Writes content to the hosts file (requires elevated permissions)
+ */
+export const writeHostsFile = async ( content: string ): Promise< void > => {
+	const hostsPath = getHostsFilePath();
+	try {
+		const tempPath = path.join( tmpdir(), 'wp-studio-hosts' );
+		await writeFile( tempPath, content );
+		const command =
+			platform() === 'win32'
+				? `type ${ tempPath } > ${ hostsPath }`
+				: `cat ${ tempPath } > ${ hostsPath }`;
+		// @ts-expect-error promisify doesn't seem typed properly.
+		await sudoExec( command, {
+			name: 'WordPress Studio',
+		} );
+	} catch ( error ) {
+		console.error( 'Error writing hosts file:', error );
+		throw error;
+	}
+};
+
+/**
+ * Create a regular expression matching the hosts entry for a given domain:
+ *
+ * 	127.0.0.1 foo.wp.cloud # Port 8000 (WordPress Studio)
+ */
+function createHostsEntryPattern( domain: string ): RegExp {
+	return new RegExp( `127\\.0\\.0\\.1\\s+${ domain.replace( /\./g, '\\.' ) }(\\s|$)`, 'i' );
+}
+
+/**
+ * Adds a domain to the hosts file pointing to 127.0.0.1
+ */
+export const addDomainToHosts = async ( domain: string, port: number ): Promise< void > => {
+	try {
+		const hostsContent = await readHostsFile();
+
+		const newContent = updateStudioBlock( hostsContent, ( entries ) => {
+			const pattern = createHostsEntryPattern( domain );
+
+			// No changes if domain already present
+			if ( entries.some( ( entry ) => entry.match( pattern ) ) ) {
+				return entries;
+			}
+
+			return [ ...entries, `127.0.0.1 ${ domain } # Port ${ port } (WordPress Studio)` ];
```

Comment:
```suggestion
			return [ ...entries, `127.0.0.1 ${ domain } # Port ${ port }` ];
```
The entry is already inside a `BEGIN WordPress Studio` block. Do we really need to mention Studio on each line, too?

Reply:
I think we can remove it, yes, but I'll leave that up to Riad.

Also,

> Let's add some unit tests for src/lib/hosts-file.ts. AI is great at that, so it shouldn't be a huge lift.

You'd think so, but two very different systems both got it spectacularly wrong and cost me tons of time. Had to reset and do it myself. In hindsight I think they got confused by the use of `promisify` in `hosts-file`. :)